packages/flutter_tools/lib/src/commands/run.dart

```diff
@@ -526,6 +527,89 @@ class RunCommand extends RunCommandBase {
     );
   }
 
+  @override
+  Future<analytics.Event> unifiedAnalyticsUsageValues(String commandPath) async {
```

Comment:
it looks there are a lot of slow things here. Maybe the easiest way to solve this would be to move MOST of this logic into a `late final` expression (making it lazily evaluated in case any of this code is dependent on setup that happens before runCommand) and basically bundle it in a record:

```dart
  late final Future<AnalyticsValuesRecord> _sharedAnalyticsValues = (() async {
    final String deviceType, deviceOsVersion;
    final bool isEmulator;
    bool anyAndroidDevices = false;
    bool anyIOSDevices = false;
    bool anyWirelessIOSDevices = false;

    if (devices == null || devices!.isEmpty) {
      deviceType = 'none';
      deviceOsVersion = 'none';
      isEmulator = false;
    } else if (devices!.length == 1) {
      final Device device = devices![0];
      final TargetPlatform platform = await device.targetPlatform;
      anyAndroidDevices = platform == TargetPlatform.android;
      anyIOSDevices = platform == TargetPlatform.ios;
      if (device is IOSDevice && device.isWirelesslyConnected) {
        anyWirelessIOSDevices = true;
      }
      deviceType = getNameForTargetPlatform(platform);
      deviceOsVersion = await device.sdkNameAndVersion;
      isEmulator = await device.isLocalEmulator;
    } else {
      deviceType = 'multiple';
      deviceOsVersion = 'multiple';
      isEmulator = false;
      for (final Device device in devices!) {
        final TargetPlatform platform = await device.targetPlatform;
        anyAndroidDevices = anyAndroidDevices || (platform == TargetPlatform.android);
        anyIOSDevices = anyIOSDevices || (platform == TargetPlatform.ios);
        if (device is IOSDevice && device.isWirelesslyConnected) {
          anyWirelessIOSDevices = true;
        }
        if (anyAndroidDevices && anyIOSDevices) {
          break;
        }
      }
    }

    String? iOSInterfaceType;
    if (anyIOSDevices) {
      iOSInterfaceType = anyWirelessIOSDevices ? 'wireless' : 'usb';
    }

    String? androidEmbeddingVersion;
    final List<String> hostLanguage = <String>[];
    if (anyAndroidDevices) {
      final AndroidProject androidProject = FlutterProject.current().android;
      if (androidProject.existsSync()) {
        hostLanguage.add(androidProject.isKotlin ? 'kotlin' : 'java');
        androidEmbeddingVersion = androidProject.getEmbeddingVersion().toString().split('.').last;
      }
    }
    if (anyIOSDevices) {
      final IosProject iosProject = FlutterProject.current().ios;
      if (iosProject.exists) {
        final Iterable<File> swiftFiles = iosProject.hostAppRoot
            .listSync(recursive: true, followLinks: false)
            .whereType<File>()
            .where((File file) => globals.fs.path.extension(file.path) == '.swift');
        hostLanguage.add(swiftFiles.isNotEmpty ? 'swift' : 'objc');
      }
    }

    final BuildInfo buildInfo = await getBuildInfo();
    final String modeName = buildInfo.modeName;
    return (
      runIsEmulator: isEmulator,
      runTargetName: deviceType,
      runTargetOsVersion: deviceOsVersion,
      runModeName: modeName,
      runProjectModule: FlutterProject.current().isModule,
      runProjectHostLanguage: hostLanguage.join(','),
      runAndroidEmbeddingVersion: androidEmbeddingVersion,
      runEnableImpeller: enableImpeller.asBool,
      runIOSInterfaceType: iOSInterfaceType,
      runIsTest: targetFile.endsWith('_test.dart'),
    );
  })();
```

For convenience I would add this typedef:

```dart
typedef AnalyticsValuesRecord = ({
  String? runAndroidEmbeddingVersion,
  bool? runEnableImpeller,
  String? runIOSInterfaceType,
  bool runIsEmulator,
  bool runIsTest,
  String runModeName,
  String runProjectHostLanguage,
  bool runProjectModule,
  String runTargetName,
  String runTargetOsVersion,
});
```

Then you call reference this record from both analytics getters, and unwrap the record values and pass them to the desired constructor.

Reply:
I like this solution a lot! Great way to cache the results from the first invocation :)